internal/extractors/aggregator.go

```diff
@@ -0,0 +1,111 @@
+package extractors
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"io"
+
+	"github.com/innosat-mats/rac-extract-payload/internal/common"
+	"github.com/innosat-mats/rac-extract-payload/internal/innosat"
+)
+
+// Aggregator sorts and accumulates standalone and multi-packets
+func Aggregator(target chan<- common.DataRecord, source <-chan common.DataRecord) {
+	defer close(target)
+	var multiPackBuffer *bytes.Buffer
+	var multiPackStart common.DataRecord
+
+	for sourcePacket := range source {
+		if sourcePacket.Error != nil {
+			target <- sourcePacket
+			continue
+		}
+		reader := bytes.NewReader(sourcePacket.Buffer)
+		switch sourcePacket.SourceHeader.PacketSequenceControl.GroupingFlags() {
+		case innosat.SPStandalone:
+			// Produce error for unfinished multipack lingering
+			if multiPackBuffer != nil {
+				target <- makeUnfinishedMultiPackError(multiPackBuffer, sourcePacket)
+				multiPackBuffer = nil
+			}
+
+			// Report standalone pack
+			target <- sourcePacket
+		case innosat.SPStart:
+			// Produce error for unfinished multipack lingering
+			if multiPackBuffer != nil {
+				target <- makeUnfinishedMultiPackError(multiPackBuffer, sourcePacket)
+			}
+
+			// Start new multipack
+			multiPackBuffer = bytes.NewBuffer([]byte{})
```

Comment:
You dont't need to reallocate memory every time, use Reset() to reuse the backing array (initiate the buffer once at startup). Check for started multipack with a bool instead.

Reply:
fixed (multiPackStart is a struct so a bit uncomfortable checking against it)